// view-script/lib/evaluate.js
var deep_get = require('./deep_get'),
	parse = require('./parse'),
	iter = require('./iter'),
	err = require('./err')

module.exports = evaluate


function evaluate(expr, view) {
	if(expr === undefined) return
	var stack = expr instanceof Array ? expr.reverse() : [expr], results = []

	while(stack.length) {
		var call = stack.pop()

		if(call === undefined)
			return err("Unable to evaluate " + expr, view.comment)

		if(typeof call === 'string')
			handle_parse(call, stack, results, view)

		else if(call.hasOwnProperty('val'))
			results.push(call.val)

		else if(call.key !== undefined)
			retrieve_key(call, results, stack, view)

		else if(call.apply) apply_fn(call, results, view)

		else err("Unable to evaluate " + expr, view.comment)
	}

	if(results.length === 0) return undefined
	if(results.length === 1) return results[0]
	else return results
}


function retrieve_key(call, results, stack, view) {
	var val = deep_get(call.key, view)

	// If we're on a lazy function, push all the un-evaluated terms from the stack
	// into the results and apply the function to those
	if(val && val._lazy && typeof val._lazy === 'function') {
		var param_len = call.len - call.pos - 1
		for(var i = 0; i < param_len; ++i) {
			var param = stack.pop()
			if(param.key) {
				param.len = stack.length-1
				param.pos = i
			}
			results.push(param)
		}
		stack.push({apply: val._lazy, param_len: param_len, key_name: call.key})
	}

	// Evaluate each of the arguments to the function before applying it
	else if(typeof val === 'function') {
		var param_len = call.len - call.pos - 1
		stack.splice(stack.length - param_len, 0, {
			apply: val, param_len: call.len - call.pos - 1, key_name: call.key
		})
	}

	else if(val === undefined && call.first)
		stack.splice(0, stack.length)

	else results.push(val)
}


function apply_fn(call, results, view) {
	if(call.param_len !== undefined)
		var args = results.splice(results.length - call.param_len, call.param_len)
	else args = results.splice(0)

	if(view.parent && call.key_name.indexOf("parent"))
		var view_apply = view.parent
	else
		var view_apply = view

	results.push(call.apply.apply(view, args))
}


function handle_parse(expr, stack, results, view) {
	var sub_exprs = parse(expr, view.comment)

	sub_exprs[0].first = true

	for(var i = sub_exprs.length-1; i >= 0; --i) {
		if(sub_exprs[i].key) {
			sub_exprs[i].len = sub_exprs.length
			sub_exprs[i].pos = i
		}
		stack.push(sub_exprs[i])
	}
}
